htmresearch/algorithms/column_pooler.py

```diff
@@ -83,319 +75,389 @@ def __init__(self,
     """
     Learning mode: we are learning a new object. If there is no prior
-    activity, we randomly activate 2% of cells and create connections to
+    activity, we randomly activate 1% of cells and create connections to
     incoming input. If there was prior activity, we maintain it.
 
-    These cells will represent the object and learn distal connections to
-    lateral cortical columns.
+    These cells will represent the object and learn distal connections to each
+    other and to lateral cortical columns.
 
     Parameters:
     ----------------------------
-    @param  feedforwardInput (set)
-            Indices of active input bits
+    @param  feedforwardInput (iterable)
+            List of indices of active feedforward input bits
 
-    @param  lateralInput (set)
-            Indices of active cells from neighboring columns.
+    @param  lateralInputs (list of iterables)
+            Lists of indices of active lateral input bits, one per lateral layer
     """
-    # If there are no previously active cells, select random subset of cells
+
+    prevActiveCells = self.activeCells
+
+    # If there are no previously active cells, select random subset of cells.
+    # Else we maintain previous activity.
     if len(self.activeCells) == 0:
-      self.activeCells = set(self._random.shuffle(
-            numpy.array(range(self.numberOfCells()),
-                        dtype="uint32"))[0:self.numActiveColumnsPerInhArea])
-
-    # else we maintain previous activity, nothing to do.
-
-    # Those cells that remain active will learn on their proximal and distal
-    # dendrites as long as there is some input.  If there are no
-    # cells active, no learning happens.  This only happens in the very
-    # beginning if there has been no bottom up activity at all.
-    if len(self.activeCells) > 0:
-
-      # Learn on proximal dendrite if appropriate
-      if len(feedforwardInput) > 0:
-        self._learnProximal(feedforwardInput, self.activeCells,
-                            self.maxNewProximalSynapseCount,
-                            self.proximalPermanences,
-                            self.proximalConnections,
-                            self.initialProximalPermanence,
-                            self.synPermProximalInc, self.synPermProximalDec,
-                            self.connectedPermanence)
-
-      # Learn on distal dendrites if appropriate
-      self.tm.activateCells(
-        activeColumns=sorted(self.activeCells),
-        reinforceCandidatesExternalBasal=sorted(reinforceCandidatesExternal),
-        growthCandidatesExternalBasal=sorted(growthCandidatesExternal),
-        learn=True)
-
-
-  def _activateCellsInferenceMode(self, feedforwardInput):
+      self.activeCells = sorted(_sampleRange(self._random,
+                                             0, self.numberOfCells(),
+                                             step=1, k=self.sdrSize))
+
+    if len(feedforwardInput) > 0:
```

Comment:
I don't think so? Maybe we want to learn synapses even when there are very few input bits. Usually the minThresholdProximal is applied at the cell level, not the layer level. This is the equivalent of this line in the current implementation. https://github.com/numenta/nupic.research/blob/93ad0cf72a2262a8c1f608bba6b938262774b33b/htmresearch/algorithms/column_pooler.py#L246